syntax = "proto3";

//This is designed to be included by the main xbos proto file and includes the
//definitions for the XBOS IoT devices
//
//Maintainer: Gabe Fierro
//Version 1.0
package xbospb;

import "nullabletypes.proto";
import "weather_station.proto";

message URI {
    string namespace = 1;
    string value = 2;
}

message Triple {
    URI subject = 1;
    URI predicate = 2;
    URI object = 3;
}

enum FanMode {
    FanAuto = 0;
    FanOn = 1;
    FanOff = 2;
}

enum HVACMode {
    HVACModeOff = 0;
    HVACModeHeatOnly = 1;
    HVACModeCoolOnly = 2;
    HVACModeAuto = 3;
}

enum HVACState {
    HVACStateOff = 0;
    HVACStateHeatStage1 = 1;
    HVACStateCoolStage1 = 2;
    HVACStateHeatStage2 = 3;
    HVACStateCoolStage2 = 4;
}

enum DREventStatus {
    DREventStatusNotConfigured = 0;
    DREventStatusUnusable =  1;
    DREventStatusInactive = 2;
    DREventStatusActive = 3;
}

enum DREventType {
    DREventTypeNoEvent = 0;
    DREventTypeNormal = 1;
    DREventTypeModerate = 2;
    DREventTypeHigh = 3;
    DREventTypeSpecial = 4;
}

enum OccupancyType {
    OccupancyTypeUnknown = 0;
    OccupancyTypeOccupied = 1;
    OccupancyTypeUnoccupied = 2;
}

message XBOSIoTDeviceState {
    // current time at device/service
    // unit:ns
    uint64 time = 1;
    // unique identifier for this request; used to line up with device state requests
    int64 requestid = 2;
    // any error that occured since the last device report. If requestid above is non-zero,
    // then this error corresponds to the request with the given requestid
    string error = 3;

    // XBOS IoT devices
    Thermostat thermostat = 4;
    Meter meter = 5;
    Light light = 6;
    EVSE evse = 7;
    WeatherStation weather_station = 8;
    WeatherStationPrediction weather_station_prediction = 9;
}


message XBOSIoTDeviceActuation {
    // current time at device/service
    // unit:ns
    uint64 time = 1;
    // unique identifier for this request; used to line up with device state responses
    int64 requestid = 2;

    // XBOS IoT devices
    Thermostat thermostat = 3;
    Meter meter = 4;
    Light light = 5;
    EVSE evse = 6;
}

message XBOSIoTContext {
    // current time at device/service
    // unit:ns
    uint64 time = 1;
    // any triples this device wants to add about itself
    // these triples will be assumed to be generated by the entity
    // who has created/signed this message
    repeated Triple context = 2;
}

// Thermostat
message Thermostat {
    // Current temperature recorded by thermostat
    // unit:celsius
    Double temperature = 1;
    // unit:% rh
    Double relative_humidity = 2;
    // If true, then the thermostat is in override mode
    // unit: t/f
    Bool override = 3;
    // If true, the fan is on; else it is off
    // unit: t/f
    Bool fan_state = 4;
    // Current operating mode of the fan
    // unit: xbos/iot/FanMode
    FanMode fan_mode = 5;
    // Current operating mode of the HVAC
    // unit: xbos/iot/HVACMode
    HVACMode mode = 6;
    // Current HVAC state
    // unit: xbos/iot/HVACState
    HVACState state = 7;
    // number of heat stages enabled
    Int32 enabled_heat_stages = 8;
    // number of cool stages enabled
    Int32 enabled_cool_stages = 9;
    // heating setpoint
    // unit: celsius
    Double heating_setpoint = 10;
    // cooling setpoint
    // unit: celsius
    Double cooling_setpoint = 11;
}

message ThermostatSchedule {
    //Map day of week to daily schedule
    map<string, ThermostatScheduleDay> scheduleMap = 1;
}

message ThermostatScheduleDay {
    //Daily schedule is Multiple Blocks
    repeated ThermostatScheduleBlock blocks = 1;
}

message ThermostatScheduleBlock {
    //heating setpoint
    // unit: celsius
    Double heating_setpoint = 1;
    //cooling setpoint
    // unit: celsius
    Double cooling_setpoint = 2;
    //Current system mode of thermostat
    // unit: xbos/iot/HVACMode
    HVACMode mode = 3;
    //Time when schedule block takes effect
    //format: RRule
    string time = 4;
}

message DREvent {
    //End of History Date Range
    //unit: seconds
    Int64 event_end = 1;
    //Start of History Date Range
    //unit: seconds
    Int64 event_start = 2;
    //Current event type of DR Event
    //unit: xbos/iot/DREventType
    DREventType event_type = 3;
    //Current system status of DR event
    //unit: xbos/iot/DREventStatus
    DREventStatus dr_status = 4;
    //Time when event takes effect
    //unit: seconds
    Int64 time = 5;
}

message ThermostatOccupancy {
    //Whether Thermostat is Occupied
    Bool occupancy = 1;
    //Time when occupancy status was recorded
    //unit:seconds
    Int64 time = 2;
}

message Meter {
    // unit: kW
    Double power = 1;
    // unit: V
    Double voltage = 2;
    // unit: kVA
    Double apparent_power = 3;
    // unit: KWh
    Double energy = 4;
}

message Light {
    // True if the light is on
    // unit: on/off
    Bool state = 1;
    // 100 is maximum brightness
    Int64 brightness = 2;
}

message EVSE {
    // maximum allowed current for charging
    // unit: A
    Double current_limit = 1;
    // active charge current
    // unit: A
    Double current = 2;
    // active charge voltage
    // unit: V
    Double voltage = 3;
    // seconds left until car is charged
    // unit: seconds
    Int32 charging_time_left = 4;
    // charge state of the EVSE
    // unit: on/off
    Bool state = 5;
}
